jina/types/arrays/neuraloperations.py

```diff
@@ -0,0 +1,140 @@
+
+        Note:
+            - If metric is 'cosine' it uses the cosine **distance**.
+            - If metric is 'euclidean' it uses the euclidean **distance**.
+            - If metric is 'euclidean_squared' it uses the euclidean **distance** squared.
+
+        :param darray: the other DocumentArray to match against
+        :param metric: the distance metric
+        :param limit: the maximum number of matches, when not given
+                      all Documents in `another` are considered as matches
+        :param is_distance: Boolean flag informing if `metric` values want to be considered as distances or scores.
+        """
+
+        def invert_if_score(
+            X: 'np.ndarray', metric: str, is_distance: bool
+        ) -> 'np.ndarray':
+            """Invert values to scores if `is_distance=False` according to the metric that is passed.
+
+            :param X: input np.ndarray
+            :param metric: string defining distance function ['cosine', 'euclidean', euclidean_squared'].
+            :param is_distance: Boolean flag that describes if values in X need to be reinterpreted as similarities.
+            :return: A np.ndarray with distances if `is_distance=True`, scores if distance=False`.
+            """
+            if metric == 'cosine':
+                if is_distance:
+                    X = X
+                else:
+                    X = 1 - X
+            if metric == 'euclidean_squared' or metric == 'euclidean':
+                if is_distance:
+                    X = X
+                else:
+                    X = 1 / (X + 1)
+            return X
+
+        X = np.stack(self.get_attributes('embedding'))
+        Y = np.stack(darray.get_attributes('embedding'))
+        limit = min(limit, len(darray))
+
+        dists = compute_distances(X, Y, metric)
+        idx, dist = self._get_sorted_smallest_k(dists, limit)
+
+        for _q, _ids, _dists in zip(self, idx, dist):
+            for _id, _dist in zip(_ids, _dists):
+                d = Document(darray[int(_id)], copy=True)
+                d.scores[metric] = invert_if_score(_dist, metric, is_distance)
+                _q.matches.append(d)
+
+    @staticmethod
+    def _get_sorted_smallest_k(
+        dists: 'np.ndarray', top_k: int
+    ) -> Tuple['np.ndarray', 'np.ndarray']:
+        """Finds the smallest `top_k` values (and its indices) from `dists`. Returns both indices and values.
+
+        :param dists: np.ndarray of distances
+        :param top_k: number of values to retrieve
+        :return: Lowest values in dists and the indices of those values
+        """
+        if top_k >= dists.shape[1]:
+            indices = dists.argsort(axis=1)[:, :top_k]
+            dists = np.take_along_axis(dists, indices, axis=1)
+        else:
+            indices_ps = dists.argpartition(kth=top_k, axis=1)[:, :top_k]
+            dists = np.take_along_axis(dists, indices_ps, axis=1)
+            indices_fs = dists.argsort(axis=1)
```

Comment:
Can you explain what is the point of this? You are doing partial sort first, and full sort later- which means you are loosing any performance gain that partial sort would bring. 

I would either just use the full sort, or re-write this to use partial sort only (if it's possible).

Reply:
Ah never mind, I see you change `dists`.